cpp/path_module/algorithm/path.cpp

```diff
@@ -0,0 +1,173 @@
+#include "path.hpp"
+
+void GetStartNodes(const mgp::Value element, const mgp::Graph &graph, std::unordered_set<mgp::Node> &start_nodes) {
+  if (!(element.IsNode() || element.IsInt())) {
+    throw mgp::ValueException("First argument must be type node, id or a list of those.");
+  }
+  if (element.IsNode()) {
+    start_nodes.insert(element.ValueNode());
+    return;
+  }
+  start_nodes.insert(graph.GetNodeById(mgp::Id::FromInt(element.ValueInt())));
+}
+
+bool RelFilterAllows(const mgp::Map &config, std::string_view type, bool ingoing) {
+  mgp::List list_of_types = config.At("relationshipFilter").ValueList();
+  if (list_of_types.Size() == 0) {
+    return true;
+  }
+  for (const auto element : list_of_types) {
+    auto string_rel_type = element.ValueString();
+    if (string_rel_type.front() == '<' && ingoing) {
+      string_rel_type.remove_prefix(1);
+    }
+    if (string_rel_type.back() == '>' && !ingoing) {
+      string_rel_type.remove_suffix(1);
+    }
+    if (string_rel_type == type || string_rel_type.size() == 0) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool IsLabelListed(const mgp::Node node, std::unordered_set<std::string_view> &set) {
+  for (const auto label : node.Labels()) {
+    if (set.contains(label)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+void Path::VisitNode(const mgp::Node node, std::unordered_map<mgp::Node, std::int64_t> &visited_nodes, bool is_start,
+                     const mgp::Map &config, int64_t hop_count, Path::LabelSets &labelFilterSets,
+                     const mgp::RecordFactory &record_factory) {
+  if (config.At("maxLevel").ValueInt() != -1 && hop_count > config.At("maxLevel").ValueInt()) {
+    return;
+  }
+  if (config.At("filterStartNode").ValueBool() || !is_start) {
+    if ((IsLabelListed(node, labelFilterSets.blacklist)) ||
+        (!labelFilterSets.whitelist.empty() && !IsLabelListed(node, labelFilterSets.whitelist) &&
+         !IsLabelListed(node, labelFilterSets.end_list) && !IsLabelListed(node, labelFilterSets.termination_list))) {
+      return;
+    }
+  }
```

Comment:
I know these could be united but I think this way it is easier to follow what's going on?

Reply:
Makes sense to me